## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [36]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [37]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [38]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [39]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.tail()



first       last                      job               phone  \
54634   Lauren      Reese           Radio producer  977.877.5272x11382   
54635  Derrick      Moore                 Musician        358.661.5483   
54636   Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637    Wanda    Solomon             Toxicologist       (311)767-4924   
54638   Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                    email  client_id  order_id  order_date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                    belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5

In [40]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [41]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [42]:
# For the category with the most entries,
# which subcategory had the most entries?
df.loc[df['category'] == 'consumables', 'subcategory'].value_counts().head(1)

#df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [43]:
# Which five clients had the most entries in the data?

df['client_id'].value_counts().head()



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [44]:
# Store the client ids of those top 5 clients in a list.

top_5_clients = df["client_id"].value_counts().head().index.tolist()

print(top_5_clients)


[33615, 66037, 46820, 38378, 24741]


In [45]:
# How many total units (the qty column) did the
# client with the most entries order order?
#df.groupby('client_id')['qty'].value_counts().sum()

df.loc[df['client_id'] == 33615, 'qty'].sum()


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [46]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty


line_subtotal = df['unit_price']*df['qty']
df["line_subtotal"] = line_subtotal



df[["unit_price","qty", "line_subtotal","client_id"]].head()



unit_price  qty  line_subtotal  client_id
0     1096.80  105      115164.00      58515
1       24.95   21         523.95      37609
2       13.52   39         527.28      57113
3       36.42   29        1056.18      46554
4      195.10   20        3902.00      92089

In [47]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
total_weight = df['unit_weight']*df['qty']
df["total_weight"] = total_weight
def shipping(total_weight):
    if total_weight < 50:
        return total_weight * 10
    return total_weight * 7

df['Shipping_price'] = df['total_weight'].apply(shipping)
df[["unit_price","qty","unit_weight", "total_weight","Shipping_price"]].head()

unit_price  qty  unit_weight  total_weight  Shipping_price
0     1096.80  105         7.50        787.50         5512.50
1       24.95   21         1.49         31.29          312.90
2       13.52   39         1.68         65.52          458.64
3       36.42   29         1.23         35.67          356.70
4      195.10   20        46.43        928.60         6500.20

In [48]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

line_price = (df['Shipping_price'] + df['line_subtotal']) * 1.0925
df["line_price"] = line_price

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  total_weight  Shipping_price  \
0         7.50  105            1      115164.00        787.50         5512.50   
1         1.49   21            0         523.95         31.29          312.90   
2         1.68   39            6         527.28         65.52          458.64   
3         1.23   29            3        1056.18         35.67          356.70   
4        46.43   20            1        3902.00        928.60         6500.20   

      line_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [49]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

line_cost = df['unit_cost'] * df['qty'] + df['Shipping_price'] 
df["line_cost"] = line_cost

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  Shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   
3            3        1056.18         35.67          356.70    1543.571400   
4            1        3902.00        928.60         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [50]:
# Create a column for the profit of
# each line using line cost and line price

line_profit =  df['line_price'] - df['line_cost']
df["line_profit"] = line_profit

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  Shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   
3        1056.18         35.67          356.70    1543.571400    1077.35   
4        3902.00        928.60         6500.20   11364.403500    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [51]:
# Check your work using the totals above
total_orders_df = df.groupby('order_id').agg({
        'line_price': 'sum',
})
order_ids = [2742071,2173913,6128929]
#three_orders_df=[df['order_id'].apply(lambda x: x in orders_id )] 

total_orders_df[["line_price"]]

def find_order_price(order_id):
    order_lines_df = df.loc[df['order_id'] == order_id, 'line_price']
    return round(order_lines_df.sum(), 2)

for order_id in order_ids:
    print(f"Order {order_id} total: ${find_order_price(order_id)}")


Order 2742071 total: $152811.9
Order 2173913 total: $162388.72
Order 6128929 total: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [52]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

filtered_df = df[df['client_id'].apply(lambda x: x in top_5_clients)]

#filtered_df[["client_id", "unit_price","qty", "line_subtotal"]].head()

filtered_df.head()

first     last                                           job  \
5    Jessica    Reyes  Production designer, theatre/television/film   
12   Jessica    Reyes  Production designer, theatre/television/film   
137    Bryan    Myers                                Chief of Staff   
214    Bryan    Myers                                Chief of Staff   
216   Angela  Everett                                     Osteopath   

                phone                     email  client_id  order_id  \
5          5016915923  jessica.reyes@grimes.net      33615   1327730   
12         5016915923  jessica.reyes@grimes.net      33615   6059361   
137  444.233.9146x233     bmyers3696@wilson.com      46820   7109502   
214  444.233.9146x233     bmyers3696@wilson.com      46820   8967498   
216  729-620-6192x875    aeverett@rodriguez.com      66037   1203693   

     order_date  order_week  order_year  ... unit_cost unit_weight  qty  \
5    2023-05-02          18        2023  ...    154.44        2.79   78   
12   2023-05-16          20        2023  ...    198.38       27.51  568   
137  2023-05-12          19        2023  ...    111.47        5.65  118   
214  2023-02-06           6        2023  ...      6.26        7.05  133   
216  2023-04-09          14        2023  ...    166.09        0.26  157   

     line_number  line_subtotal  total_weight  Shipping_price     line_price  \
5              0       13481.52        217.62         1523.34   16392.809550   
12             2      130492.32      15625.68       109379.76  262060.247400   
137            9       16063.34        666.70         4666.90   22647.787200   
214            3        1061.34        937.65         6563.55    8330.192325   
216            2       47393.59         40.82          408.20   52223.455575   

     line_cost   line_profit  
5     13569.66   2823.149550  
12   222059.60  40000.647400  
137   17820.36   4827.427200  
214    7396.13    934.062325  
216   26484.33  25739.125575  

[5 rows x 24 columns]

In [53]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
total_df = filtered_df.groupby('client_id').agg({
    'qty': 'sum',
    'Shipping_price': 'sum',
    'line_price':'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
})
summary_df = total_df.sort_values(by='line_profit', ascending=False)

summary_df.head()

qty  Shipping_price    line_price    line_cost   line_profit
client_id                                                                 
24741      239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07
38378       73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
66037       43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
46820       75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
33615       64313      1828834.89  8.377145e+06   6175163.91  2.201981e+06

In [54]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
summary_df = summary_df.reset_index()

client_df = summary_df.rename(columns={"client_id":"Client ID",
                                        "qty": "Quantity",
                                        "shipping_price": "Shipping",
                                        "line_price": "Total Revenue",
                                        "line_cost": "Total Cost",
                                        "line_profit": "Total Profit", })

#def scale_to_millions(column):

    #return '{:,.2f}M'.format(column/1000000)

#client_df['Total Price'] = client_df.apply(scale_to_millions, axis = 1)

client_df.head()

Client ID  Quantity  Shipping_price  Total Revenue   Total Cost  \
0      24741    239862      5126448.37   8.226889e+07  45688899.71   
1      38378     73667      3429455.40   1.290655e+07   9634720.98   
2      66037     43018      1395151.85   1.025951e+07   7004482.98   
3      46820     75768      1601448.84   9.743794e+06   7007191.64   
4      33615     64313      1828834.89   8.377145e+06   6175163.91   

   Total Profit  
0  3.657999e+07  
1  3.271830e+06  
2  3.255032e+06  
3  2.736603e+06  
4  2.201981e+06

In [55]:
# Sort the updated data by "Total Profit" form highest to lowest
client_df.sort_values(by='Total Profit', ascending=False)

Client ID  Quantity  Shipping_price  Total Revenue   Total Cost  \
0      24741    239862      5126448.37   8.226889e+07  45688899.71   
1      38378     73667      3429455.40   1.290655e+07   9634720.98   
2      66037     43018      1395151.85   1.025951e+07   7004482.98   
3      46820     75768      1601448.84   9.743794e+06   7007191.64   
4      33615     64313      1828834.89   8.377145e+06   6175163.91   

   Total Profit  
0  3.657999e+07  
1  3.271830e+06  
2  3.255032e+06  
3  2.736603e+06  
4  2.201981e+06